### PR TITLE
[SPARK-41986][SQL] Introduce shuffle on SinglePartition

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR introduces shuffle on SinglePartition if it's physical size greater than `spark.sql.maxSinglePartitionBytes`.

### Why are the changes needed?

Improve parallelism. For example:
```scala
spark.range(100000000L).selectExpr("id as a", "id as b").write.saveAsTable("t1")

sql(
  """
    |WITH base
    |     AS (select *, ROW_NUMBER() OVER(ORDER BY a) AS new_a FROM t1)
    |SELECT * FROM base t1 JOIN base t2 ON t1.a = t2.b
    |""".stripMargin).explain()
```

Before this PR:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [a#10L], [b#26L], Inner
   :- Filter isnotnull(a#10L)
   :  +- Window [row_number() windowspecdefinition(a#10L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS new_a#8], [a#10L ASC NULLS FIRST]
   :     +- Sort [a#10L ASC NULLS FIRST], false, 0
   :        +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=50]
   :           +- FileScan parquet spark_catalog.default.t1[a#10L,b#11L]
   +- Sort [b#26L ASC NULLS FIRST], false, 0
      +- Filter isnotnull(b#26L)
         +- Window [row_number() windowspecdefinition(a#25L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS new_a#27], [a#25L ASC NULLS FIRST]
            +- Sort [a#25L ASC NULLS FIRST], false, 0
               +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=54]
                  +- FileScan parquet spark_catalog.default.t1[a#25L,b#26L]
```

After this PR:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [a#10L], [b#26L], Inner
   :- Sort [a#10L ASC NULLS FIRST], false, 0
   :  +- Exchange hashpartitioning(a#10L, 5), ENSURE_REQUIREMENTS, [plan_id=60]
   :     +- Filter isnotnull(a#10L)
   :        +- Window [row_number() windowspecdefinition(a#10L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS new_a#8], [a#10L ASC NULLS FIRST]
   :           +- Sort [a#10L ASC NULLS FIRST], false, 0
   :              +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=50]
   :                 +- FileScan parquet spark_catalog.default.t1[a#10L,b#11L]
   +- Sort [b#26L ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(b#26L, 5), ENSURE_REQUIREMENTS, [plan_id=61]
         +- Filter isnotnull(b#26L)
            +- Window [row_number() windowspecdefinition(a#25L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS new_a#27], [a#25L ASC NULLS FIRST]
               +- Sort [a#25L ASC NULLS FIRST], false, 0
                  +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=54]
                     +- FileScan parquet spark_catalog.default.t1[a#25L,b#26L]
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan @sunchao

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:84]: Looks OK, but I'm not totally sure whether it's good to reuse `SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES` here. It is for AQE and this code is used by both AQE and non-AQE?

Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:84]: How about `SQLConf.FILES_MAX_PARTITION_BYTES`?

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala:544]: is it due to wrong size estimation?

Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala:544]: The size is defaultSizeInBytes:
https://github.com/apache/spark/blob/f24f8489f8096c5324a6e7084437ee2238311103/sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala#L330

https://github.com/apache/spark/blob/074e1b39d279f12ff8d822a03741f33f159f5df8/sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala#L53-L57

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: Added a new config: spark.sql.maxSinglePartitionBytes

**Review by sunchao — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:522]: For the safety, can we use more higher value, @wangyum ?

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:522]: I'd recommend to use `Long.MaxValue` like `spark.sql.defaultSizeInBytes` in Apache Spark 3.4.0. We can reduce it in 3.5.0 and later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:517]: Also, shall we make this as a user-facing conf?

**Review by dongjoon-hyun — COMMENTED** [review]
I have two comments on the config visibility and the default value in Apache Spark 3.4.0.

cc @xinrong-meng as Apache Spark 3.4.0 release manager.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs). Thank you so much for the update, @wangyum .

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.4.0.

**Comment by cloud-fan** [comment]
@wangyum can we update the PR description? It does not match the real cod changes which is confusing.

**Comment by wangyum** [comment]
@cloud-fan Has updated it.

**Pull request closed without merging**
